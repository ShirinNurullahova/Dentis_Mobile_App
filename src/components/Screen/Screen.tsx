import React, {FC, useEffect, useRef, useState} from 'react';
import {
  Keyboard,
  KeyboardAvoidingView,
  Platform,
  RefreshControl,
  ScrollView,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
  ViewStyle,
} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';
import {useDispatch, useSelector} from 'react-redux';
import {changeScroll} from '../../store/reducers/smoothScrollReducer';

export type ScreenProps = {
  header?: React.ReactNode;
  onPullToRefresh?: (finishCallback: () => void) => void;
  scroll?: boolean;
  children?: JSX.Element | JSX.Element[] | string;
  notTouchable?: boolean;
  style?:ViewStyle
};

const Content: FC<ScreenProps> = ({header, children,style}) => {
  return (
    <View style={[style]}>
      {header}
      {children}
    </View>
  );
};

const Screen: FC<ScreenProps> = ({
  header,
  children,
  onPullToRefresh,
  notTouchable,
  style
}) => {
  const [keyboardAvoidingEnabled, setKeyboardAvoidingEnabled] = useState(true);
  const [refreshing, setRefreshing] = useState(false);
  const startRefreshing = () => {
    setRefreshing(true);
    onPullToRefresh && onPullToRefresh(stopRefreshing);
  };
  const stopRefreshing = () => {
    setRefreshing(false);
  };

  const handleKeyboardAvoidingViewBehavior = () => {
    if (Platform.OS === 'ios') {
      return 'padding';
    }
    return undefined;
  };
  const scrollY = useSelector(
    (state: {smoothScroll: {scrollY: number}}) => state?.smoothScroll?.scrollY,
  );
  const scrollViewRef = useRef(null);
  const dispatch = useDispatch();
  useEffect(() => {
    console.log({scrollY});
    if (scrollY) {
      scrollViewRef?.current?.scrollTo({y: scrollY, animated: true});
      dispatch(changeScroll(null));
    }
  }, [scrollY]);

  return (
    <KeyboardAvoidingView
      id={handleKeyboardAvoidingViewBehavior()}
      testID="keyboardAvoidingView"
      behavior={handleKeyboardAvoidingViewBehavior()}
      enabled={keyboardAvoidingEnabled}>
      <ScrollView
        ref={scrollViewRef}
        contentContainerStyle={{flexGrow: 1}}
        testID="refreshControl"
        refreshControl={
          onPullToRefresh && (
            <RefreshControl
              refreshing={refreshing}
              onRefresh={startRefreshing}
            />
          )
        }
        keyboardShouldPersistTaps={'always'}>
        {notTouchable ? (
          <SafeAreaView style={{flex:1}}>
            <Content header={header} children={children} style={style || styles.container}/>
          </SafeAreaView>
        ) : (
          <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
            <SafeAreaView>
              <Content header={header} children={children} style={style || styles.container}/>
            </SafeAreaView>
          </TouchableWithoutFeedback>
        )}
      </ScrollView>
    </KeyboardAvoidingView>
  );
};

const styles = StyleSheet.create({
  container:{
    paddingHorizontal: 16,
    paddingVertical:100
  }
})

export default React.memo(Screen);
