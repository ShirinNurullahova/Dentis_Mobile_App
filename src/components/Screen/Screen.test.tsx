import React from "react";
import { fireEvent, render } from "@testing-library/react-native";
import Screen from "./Screen";
import { Platform } from "react-native";

describe("Screen test", () => {
  test("should match with snapshot", () => {
    const comp = render(<Screen />);
    expect(comp).toMatchSnapshot();
  });
  test("Screen should call onPullToRefresh when refreshing", () => {
    const onPullToRefreshMock = jest.fn();
    const comp = render(<Screen onPullToRefresh={onPullToRefreshMock} />);
    const refreshControl = comp.getByTestId("refreshControl");
    fireEvent(refreshControl.props.refreshControl, "onRefresh");
    expect(refreshControl.props.refreshControl).toBeTruthy();
    expect(onPullToRefreshMock).toHaveBeenCalledTimes(1);
  });
  test("handleKeyboardAvoidingViewBehavior should return correct behavior for ios", () => {
    Platform.OS = "ios";
    const onPullToRefreshMock = jest.fn();
    const comp = render(<Screen onPullToRefresh={onPullToRefreshMock} />);
    const keyboardAvoidingView = comp.getByTestId("keyboardAvoidingView");
    fireEvent(keyboardAvoidingView, "behavior");
    expect(keyboardAvoidingView.props.id).toBe("padding");
  });
  test("handleKeyboardAvoidingViewBehavior should return correct behavior for android", () => {
    Platform.OS = "android";
    const onPullToRefreshMock = jest.fn();
    const comp = render(<Screen onPullToRefresh={onPullToRefreshMock} />);
    const keyboardAvoidingView = comp.getByTestId("keyboardAvoidingView");
    fireEvent(keyboardAvoidingView, "behavior");
    expect(keyboardAvoidingView.props.id).toBeUndefined();
  });
  //   test("should set refreshing to false", () => {
  //     const stopRefreshing = jest.fn();
  //     const onPullToRefreshMock = jest.fn(stopRefreshing);
  //     const comp = render(<Screen onPullToRefresh={onPullToRefreshMock} />);
  //     const refreshControl = comp.getByTestId("refreshControl");
  //     fireEvent(refreshControl.props.refreshControl, "onRefresh", stopRefreshing);
  //     expect(refreshControl.props.refreshing).toBeFalsy();
  //     fireEvent(comp.getByTestId("content"), "stopRefreshing");
  //     expect(stopRefreshing).toHaveBeenCalledTimes(1);
  //   });
});
