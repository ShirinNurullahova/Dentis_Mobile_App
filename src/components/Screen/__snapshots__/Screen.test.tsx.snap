// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen test should match with snapshot 1`] = `
<View
  id="padding"
  onLayout={[Function]}
  style={
    {
      "paddingBottom": 0,
    }
  }
  testID="keyboardAvoidingView"
>
  <RCTScrollView
    keyboardShouldPersistTaps="always"
    testID="refreshControl"
  >
    <View>
      <RNCSafeAreaView
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={true}
        edges={
          [
            "bottom",
            "left",
            "right",
            "top",
          ]
        }
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View />
      </RNCSafeAreaView>
    </View>
  </RCTScrollView>
</View>
`;
